# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [8]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [9]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [10]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [11]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [12]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [13]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [17]:
df['Season'].nunique()


32

Conclusion: By using the nunique() function, i have found that there are 32 unique seasons.

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [21]:
df['Wteam'].value_counts()
value_counts.max()


819

Conclusion: Here I have found the teams with the most wins, which is 819.

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [22]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [23]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [24]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [25]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [26]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [27]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [28]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [29]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [30]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [31]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [33]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Winning Team', 'Winning Score'

# Sort DataFrame by season
df_sorted_season = df.sort_values(by='Season')

# Sort DataFrame by winning team
df_sorted_team = df.sort_values(by='Wteam')

# Sort DataFrame by winning score
df_sorted_score = df.sort_values(by='Wscore')

# Select rows from index 100 to 200 and specific columns
selected_data = df.iloc[100:201, [0, 1, 2]]  # Replace [0, 1, 2] with the respective column indexes

print("Sorted DataFrame by season:")
print(df_sorted_season)

print("\nSorted DataFrame by winning team:")
print(df_sorted_team)

print("\nSorted DataFrame by winning score:")
print(df_sorted_score)

print("\nSelected rows and columns:")
print(selected_data)


Sorted DataFrame by season:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
2484      1985      99   1275      73   1132      63    H      0
2485      1985      99   1306      62   1171      59    H      0
2486      1985      99   1318      99   1145      91    H      0
2487      1985      99   1337      75   1109      56    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
141706    2016      50   1300      57   1183      52    N      0
141705    2016      50   1298      62   1202      49    H      0
141704    2016      50   1295      68   1410      50    N      0
141702    2016      50   1277      99   1324      93    N      1
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

Sorted DataFrame by winning team:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0


Conclusion:  Here I have made 3 dataframes that are sorted by season, winning team, and winning score. Then, Using iloc, i have selected the rows from index 100 to 200 and the columns for season, winning team, and winning score. 

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [38]:
# Write your code here
import pandas as pd

# Assuming you have three sorted DataFrames: df_sorted_season, df_sorted_team, df_sorted_score

# Find the game with the highest winning score from the sorted DataFrame by winning score
highest_score_row = df_sorted_score['Wscore'].idxmax()
highest_score_season = df_sorted_score.loc[highest_score_row, 'Season']
highest_score_winning_team = df_sorted_score.loc[highest_score_row, 'Wteam']

print("Season of the game with the highest winning score:", highest_score_season)
print("Winning team of the game with the highest winning score:", highest_score_winning_team)


Season of the game with the highest winning score: 1991
Winning team of the game with the highest winning score: 1258


Conclusion: I have here found the season and winning team for the game, with the highest winning score. The season of the game with the highest winning score is 1991, and the winning team of the game with the highest winning score is 1258.

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [36]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [37]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [39]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Score', 'Losing Score'

# Create the 'ScoreDifference' column
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average 'ScoreDifference'
average_difference = df['ScoreDifference'].mean()

# Filter the DataFrame based on the condition
filtered_df = df[df['ScoreDifference'] > average_difference]

print("Filtered DataFrame:")
print(filtered_df)


Filtered DataFrame:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...    

Conclusion: Here i have found the difference between the winning score and the losing score, where i have filtered the dataframe to only include games, where the score difference is greater than the average score difference for all games.

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [40]:
# Write your code here
import pandas as pd

# Assuming you have a filtered DataFrame named 'filtered_df' with columns: 'Season', 'Winning Team', 'Losing Team', 'ScoreDifference'

# Find the game with the highest 'ScoreDifference' from the filtered DataFrame
highest_difference_row = filtered_df['ScoreDifference'].idxmax()
highest_difference_season = filtered_df.loc[highest_difference_row, 'Season']
winning_team = filtered_df.loc[highest_difference_row, 'Wteam']
losing_team = filtered_df.loc[highest_difference_row, 'Lteam']

print("Season of the game with the highest 'ScoreDifference':", highest_difference_season)
print("Winning team of the game with the highest 'ScoreDifference':", winning_team)
print("Losing team of the game with the highest 'ScoreDifference':", losing_team)


Season of the game with the highest 'ScoreDifference': 1996
Winning team of the game with the highest 'ScoreDifference': 1409
Losing team of the game with the highest 'ScoreDifference': 1341


Conclusion: Now I am able to find the season and teams involved in the game with the highest score difference. Here we can se that the season of the game with the highest score difference is 1996, the winning team of the game with the highest score difference is 1409 and the losing team of the game with the highest score difference is 1341.

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [41]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [42]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [43]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [44]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [45]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Winning Score'

# Group the DataFrame by season and calculate the average winning score
average_winning_score = df.groupby('Season')['Wscore'].mean()

print("Average winning score for each season:")
print(average_winning_score)


Average winning score for each season:
Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


Conclusion: In this task I have created to rows, where one represents the season and the other average winning score. For example in season 2006 the average winning score for the season was 74.587555.

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [46]:
# Write your code here
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Team', 'Winning Score'

# Group the DataFrame by winning team and find the maximum winning score for each team
max_winning_score = df.groupby('Wteam')['Wscore'].max()

print("Maximum winning score for each team across all seasons:")
print(max_winning_score)


Maximum winning score for each team across all seasons:
Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


Conclusion: Above I have created two rows, where the first show the winning team and the maximum winning score. For exaple wteam 1460 had a maximum wining score of 136.

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [47]:
# Write your code here
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Winning Team', 'Winning Score'

# Group the DataFrame by both season and winning team and find the team with the highest average winning score for each season
highest_avg_winning_score = df.groupby(['Season', 'Wteam'])['Wscore'].agg(['mean']).idxmax()

print("Team with the highest average winning score for each season:")
print(highest_avg_winning_score)


Team with the highest average winning score for each season:
mean    (1990, 1258)
dtype: object


Conclusion: Above i have found the team with the highest average winning score.

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [48]:
import pandas as pd

# Group the DataFrame by 'Season' and 'Wteam', count the occurrences, and rename the column
win_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Print the number of wins for each team in each season
print("Number of wins for each team in each season:")
for _, row in win_count.iterrows():
    print(f"Season {row['Season']}: Team {row['Wteam']} - {row['Wins']} wins")


Number of wins for each team in each season:
Season 1985: Team 1102 - 5 wins
Season 1985: Team 1103 - 9 wins
Season 1985: Team 1104 - 21 wins
Season 1985: Team 1106 - 10 wins
Season 1985: Team 1108 - 19 wins
Season 1985: Team 1109 - 1 wins
Season 1985: Team 1110 - 7 wins
Season 1985: Team 1111 - 10 wins
Season 1985: Team 1112 - 18 wins
Season 1985: Team 1113 - 11 wins
Season 1985: Team 1114 - 17 wins
Season 1985: Team 1116 - 21 wins
Season 1985: Team 1117 - 11 wins
Season 1985: Team 1119 - 15 wins
Season 1985: Team 1120 - 18 wins
Season 1985: Team 1121 - 7 wins
Season 1985: Team 1122 - 6 wins
Season 1985: Team 1123 - 13 wins
Season 1985: Team 1124 - 7 wins
Season 1985: Team 1126 - 6 wins
Season 1985: Team 1129 - 12 wins
Season 1985: Team 1130 - 16 wins
Season 1985: Team 1131 - 14 wins
Season 1985: Team 1132 - 11 wins
Season 1985: Team 1133 - 16 wins
Season 1985: Team 1134 - 11 wins
Season 1985: Team 1135 - 8 wins
Season 1985: Team 1137 - 15 wins
Season 1985: Team 1139 - 16 wins
Season 

Conclusion: Above I have found the number of wins for each team in each season, and grouping by both season and winning team. For example in season 2016 team 1462 had 27 wins.

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [49]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Wteam'

# Group the DataFrame by both season and winning team and count the number of wins
win_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Group the win_count DataFrame by season and find the team with the most wins for each season
team_with_most_wins = win_count.groupby('Season')['Wins'].idxmax()
most_wins_df = win_count.loc[team_with_most_wins]

print("Team with the most wins for each season:")
for _, row in most_wins_df.iterrows():
    print(f"Season {row['Season']}: Team {row['Wteam']} - {row['Wins']} wins")


Team with the most wins for each season:
Season 1985: Team 1385 - 27 wins
Season 1986: Team 1181 - 32 wins
Season 1987: Team 1424 - 33 wins
Season 1988: Team 1112 - 31 wins
Season 1989: Team 1328 - 28 wins
Season 1990: Team 1247 - 29 wins
Season 1991: Team 1116 - 30 wins
Season 1992: Team 1181 - 28 wins
Season 1993: Team 1231 - 28 wins
Season 1994: Team 1163 - 27 wins
Season 1995: Team 1116 - 26 wins
Season 1996: Team 1269 - 31 wins
Season 1997: Team 1242 - 31 wins
Season 1998: Team 1242 - 33 wins
Season 1999: Team 1181 - 32 wins
Season 2000: Team 1409 - 29 wins
Season 2001: Team 1181 - 29 wins
Season 2002: Team 1153 - 30 wins
Season 2003: Team 1166 - 29 wins
Season 2004: Team 1390 - 29 wins
Season 2005: Team 1228 - 32 wins
Season 2006: Team 1181 - 30 wins
Season 2007: Team 1242 - 30 wins
Season 2008: Team 1272 - 33 wins
Season 2009: Team 1272 - 31 wins
Season 2010: Team 1242 - 32 wins
Season 2011: Team 1242 - 32 wins
Season 2012: Team 1246 - 32 wins
Season 2013: Team 1211 - 30 wins
Se

Conclusion: Above I have done the same as the task 5, but I have also used the idxmax() function. The anserws are the same as the task before.

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [77]:
# Write your code here
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Team', 'Winning Score', 'Losing Team', 'Losing Score'

# Group the DataFrame by losing team and find the average losing score for each team across all seasons
average_losing_score = df.groupby('Lteam')['Lscore'].mean()

# Retrieve the average winning score for each team from task 3
average_winning_score = df.groupby('Wteam')['Wscore'].mean()

# Compare the average losing score with the average winning score
teams_with_higher_losing_score = average_losing_score[average_losing_score > average_winning_score]

print("Teams with a higher average losing score than winning score:")
print(teams_with_higher_losing_score)


Teams with a higher average losing score than winning score:
Series([], Name: Lscore, dtype: float64)


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [53]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [55]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Score'

# Define a function to check if the winning score is greater than 100
def check_high_scoring(score):
    if score > 100:
        return 'Yes'
    else:
        return 'No'

# Create a new column 'HighScoringGame' by applying the function to 'Wscore'
df['HighScoringGame'] = df['Wscore'].apply(check_high_scoring)

print(df)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3              

Conclusion: Above I have created a new column named 'HighScoringGame' where is says 'Yes' if the winning score is greater than 100 and 'No' if it is less than 100.

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [58]:
# Write your code here
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Team', 'Losing Team'

# Create an empty dictionary to keep track of the number of games for each team
games_played = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Get the winning team and update the games_played dictionary
    winning_team = row['Wteam']
    games_played[winning_team] = games_played.get(winning_team, 0) + 1

    # Get the losing team and update the games_played dictionary
    losing_team = row['Lteam']
    games_played[losing_team] = games_played.get(losing_team, 0) + 1

# Print the total number of games played by each team
for team, num_games in games_played.items():
    print(f"Team: {team}, Games Played: {num_games}")


Team: 1228, Games Played: 992
Team: 1328, Games Played: 968
Team: 1106, Games Played: 855
Team: 1354, Games Played: 906
Team: 1112, Games Played: 981
Team: 1223, Games Played: 363
Team: 1165, Games Played: 833
Team: 1432, Games Played: 69
Team: 1192, Games Played: 908
Team: 1447, Games Played: 903
Team: 1218, Games Played: 931
Team: 1337, Games Played: 922
Team: 1226, Games Played: 847
Team: 1242, Games Played: 993
Team: 1268, Games Played: 969
Team: 1260, Games Played: 914
Team: 1133, Games Played: 949
Team: 1305, Games Played: 922
Team: 1424, Games Played: 974
Team: 1307, Games Played: 969
Team: 1288, Games Played: 925
Team: 1344, Games Played: 951
Team: 1438, Games Played: 952
Team: 1374, Games Played: 916
Team: 1411, Games Played: 903
Team: 1412, Games Played: 962
Team: 1397, Games Played: 963
Team: 1417, Games Played: 966
Team: 1225, Games Played: 880
Team: 1116, Games Played: 980
Team: 1368, Games Played: 808
Team: 1120, Games Played: 936
Team: 1391, Games Played: 879
Team: 1135,

Conclusion: Above I have calculated the total number of games played by each team, whether they won or lost. When clicking on scrollable element more data data can be viewed.

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [61]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Winning Score', 'Losing Score'

# Create an empty dictionary to store the highest score difference for each season
highest_score_diff_per_season = {}

# Iterate over the rows of the DataFrame
for _, row in df.iterrows():
    season = row['Season']
    winning_score = row['Wscore']
    losing_score = row['Lscore']
    
    # Calculate the score difference for the current game
    score_diff = winning_score - losing_score
    
    # Update the highest score difference for the corresponding season
    if season in highest_score_diff_per_season:
        highest_score_diff_per_season[season] = max(highest_score_diff_per_season[season], score_diff)
    else:
        highest_score_diff_per_season[season] = score_diff

# Print the highest score difference for each season
for season, score_diff in highest_score_diff_per_season.items():
    print(f"Season: {season}, Highest Score Difference: {score_diff}")


Season: 1985, Highest Score Difference: 60
Season: 1986, Highest Score Difference: 84
Season: 1987, Highest Score Difference: 73
Season: 1988, Highest Score Difference: 68
Season: 1989, Highest Score Difference: 70
Season: 1990, Highest Score Difference: 76
Season: 1991, Highest Score Difference: 68
Season: 1992, Highest Score Difference: 82
Season: 1993, Highest Score Difference: 81
Season: 1994, Highest Score Difference: 69
Season: 1995, Highest Score Difference: 74
Season: 1996, Highest Score Difference: 91
Season: 1997, Highest Score Difference: 80
Season: 1998, Highest Score Difference: 66
Season: 1999, Highest Score Difference: 75
Season: 2000, Highest Score Difference: 75
Season: 2001, Highest Score Difference: 75
Season: 2002, Highest Score Difference: 59
Season: 2003, Highest Score Difference: 72
Season: 2004, Highest Score Difference: 70
Season: 2005, Highest Score Difference: 64
Season: 2006, Highest Score Difference: 72
Season: 2007, Highest Score Difference: 74
Season: 200

Conclusion: Above I have found the the game with the highest score difference. For example in season 2013 the t game with the highest score difference was 58.

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [62]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [64]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Winning Team', 'Losing Team'

# Combine the 'Winning Team' and 'Losing Team' columns into a single column
teams = pd.concat([df['Wteam'], df['Lteam']])

# Calculate the total number of games played by each team
games_played = teams.value_counts()

print(games_played)


1181    1013
1314    1010
1246     999
1257     994
1242     993
        ... 
1230      64
1446      48
1289      48
1118      27
1327      22
Length: 364, dtype: int64


Conclusion: Above I have calculate the total number of games played by each team, whether they won or lost, by using the value_counts() function.

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [68]:
import pandas as pd

# Assuming you have a DataFrame named 'df' with columns: 'Season', 'Winning Score', 'Losing Score'

# Calculate the score difference for each game
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Group the DataFrame by 'Season' and find the game with the highest score difference in each season
games_highest_score_diff_df = df.loc[df.groupby('Season')['ScoreDifference'].idxmax()]

print("Game with the highest score difference for each season:")
print(games_highest_score_diff_df)


Game with the highest score difference for each season:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H   

Conclusion: Above I have found the game with the highest score difference, by using vectorized subtraction and  the groupby() function and idxmax() function.

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [70]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [71]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [72]:
type(df['Wscore'])

pandas.core.series.Series

In [73]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [74]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [75]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [76]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y